firmware/library/L1_Drivers/can.hpp

```diff
@@ -1,49 +1,828 @@
+  {
+    // 00011 = Start of Frame
+    // 00010 = ID28 ... ID21
+    // 00110 = ID20 ... ID18
+    // 00100 = SRTR Bit
+    // 00101 = IDE bit
+    // 00111 = ID17 ... 13
+    // 01111 = ID12 ... ID5
+    // 01110 = ID4 ... ID0
+    // 01100 = RTR Bit
+    // 01101 = Reserved Bit 1
+    // 01001 = Reserved Bit 0
+    // 01011 = Data Length Code
+    // 01010 = Data Field
+    // 01000 = CRC Sequence
+    // 11000 = CRC Delimiter
+    // 11001 = Acknowledge Slot
+    // 11011 = Acknowledge Delimiter
+    // 11010 = End of Frame
+    // 10010 = Intermission
+
+    interrupts[controller_].ICR = can_registers[controller_]->ICR;
+    return interrupts[controller_].flags.error_code_location;
+  }
+
+  void EnableBus() override
+  {
+    SetControllerMode(kReset, false);
+  }
+
+  [[gnu::always_inline]] static void ProcessIrq()
+  {
+    TxMessage_t * message_ptr;
+    bool transmit_interrupt, receive_interrupt;
+
+    for (uint8_t controller = 0; controller < kNumberOfControllers;
+         controller++)
+    {
+      if (is_controller_initialized[controller] == true)
+      {
+        // Capture interrupt register
+        // Bits are cleared when you read the ICR register
+        interrupts[controller].ICR = can_registers[controller]->ICR;
+
+        transmit_interrupt = interrupts[controller].flags.tx_buffer_1_ready ||
+                             interrupts[controller].flags.tx_buffer_2_ready ||
+                             interrupts[controller].flags.tx_buffer_3_ready;
+
+        receive_interrupt = interrupts[controller].flags.rx_buffer_full;
+
+        if (transmit_interrupt == true)
+        {
+          if (xQueueIsQueueEmptyFromISR(transmit_queue[controller]) == pdFALSE)
+          {
+            // dequeue message pointer
+            xQueueReceiveFromISR(transmit_queue[controller], &message_ptr, 0);
+            // send it to the available buffer
+            if (interrupts[controller].flags.tx_buffer_1_ready)
+            {
+              WriteMessageToBuffer(message_ptr, kBuffer1,
+                                   static_cast<Controllers>(controller));
+            }
+            else if (interrupts[controller].flags.tx_buffer_2_ready)
+            {
+              WriteMessageToBuffer(message_ptr, kBuffer2,
+                                   static_cast<Controllers>(controller));
+            }
+            else if (interrupts[controller].flags.tx_buffer_3_ready)
+            {
+              WriteMessageToBuffer(message_ptr, kBuffer3,
+                                   static_cast<Controllers>(controller));
+            }
+            else
+            {
+              // will never occur
+            }
+          }
+        }
+
+        if (receive_interrupt == true)
+        {
+          ReadMessageFromBuffer(static_cast<Controllers>(controller));
+        }
+      }
+    }
+  }
+
+  static void CanIrqHandler()
+  {
+    ProcessIrq();
+  }
+
+  template <typename T>
+  void CreateStaticQueues(T static_queue_config)
```

Comment:
Why is this templated?

Reply:
It doesn't look like you are using this.

Reply:
I added a templated structure that holds tx/rx queue config parameters (line 267). The idea is that the user creates an instance of the templated structure with what ever queue configs they want, and they can pass it to the templated function which will go ahead and actually create the queues. 

`  Can::StaticQueueConfig_t<10,10, Can::TxMessage_t*, Can::RxMessage_t> config;
  can1.CreateStaticQueues(config);`

Reply:
So it's 2 size variables, for tx and rx queues and then the items types which is later used by CreateStaticQueues to determine the size of each element in the queue. In my example above, the third item is a pointer to a txMessage_t and the forth item is an RxMessage_t. Here, your tx size could be a lot bigger since you're first item type is a pointer to a txMessage_t (only 4 bytes) as opposed to the second item type which is a RxMessage_t (80 bytes).

Reply:
As far as the what else the user might place for the item types, there is nothing else for now. The implementation is coupled to those types, so yeah, would be better to remove those inputs to not confuse the user.

Reply:
Removed item types. User can only pass in sizes for tx and rx queues now. 

`Can::StaticQueueConfig_t<20,10>`